### PR TITLE
from-file-json-manipulation-api

### PR DESCRIPTION
**Pull request by k-rister**

### CODE REVIEW

File: crucible/from-file-json-manipulation-api/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+This example has an `oslat.json` template that the `run.sh` script
```
Comment (rafaelfolco) [inline crucible/from-file-json-manipulation-api/README.md:1]: I'd rename the dir to run-file-api to 1) keep it short 2) use 'run-file' name everywhere

File: crucible/from-file-json-manipulation-api/README.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ingests and adds information to (or deletes information from) based on
+the variables defined at the top of the `run.sh` using the "API" that
+is defined.
+
+The "API" uses `jq` to perform "well known" operations against the
+JSON template file which results in the creation of the
+`oslat.json.run` file.  This file is then submitted to crucible using
```
Comment (rafaelfolco) [inline crucible/from-file-json-manipulation-api/README.md:8]: I'd keep json files with the .json "extension" so users can edit these files with their preferred editors.

File: crucible/from-file-json-manipulation-api/oslat.json [hunk 3]
```diff
@@ -0,0 +1,78 @@
+{
```
Comment (rafaelfolco) [inline crucible/from-file-json-manipulation-api/oslat.json:1]: I'd refer to this file as oslat.template.json or something like that

File: crucible/from-file-json-manipulation-api/oslat.json.run [hunk 4]
```diff
@@ -0,0 +1,116 @@
+{
```
Comment (rafaelfolco) [inline crucible/from-file-json-manipulation-api/oslat.json.run:1]: this one I'd refer to as oslat.run.json

File: crucible/from-file-json-manipulation-api/README.md (hunk 2 above)
Comment (k-rister) [reply crucible/from-file-json-manipulation-api/README.md:8]: Yeah, I regretted doing this later on for this reason but I was too lazy to change it at the time.  I agree that we should maintain the `.json` extension going forward.

File: crucible/from-file-json-manipulation-api/run.sh [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+#stress_ng="no"
+stress_ng="yes"
+
+samples=3
+
+mode="run"
+#mode="test"
+
+from_file="oslat.json"
```
Comment (rafaelfolco) [inline crucible/from-file-json-manipulation-api/run.sh:24]: Perhaps these settings could be in a separate file under run-file-api/config and run.sh just source the config file... my point is to keep the run.sh a completely independent and reusable script from the data itself.

File: crucible/from-file-json-manipulation-api/run.sh [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    if [ $? == 0 ]; then
+	mv "${DO_JQ_FILENAME}.tmp" "${DO_JQ_FILENAME}"
+	return 0
+    else
+	return 1
+    fi
+}
+
+# change an existing run-params parameter value
+function override_run_param() {
```
Comment (rafaelfolco) [inline crucible/from-file-json-manipulation-api/run.sh:46]: just an idea for a possible improvement: add a description to each of these functions and list them all if user runs ./run.sh --help
override_run_param: change an existing run-params parameter value
...

File: crucible/from-file-json-manipulation-api/run.sh [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+# at config array index 0 for the endpoint object at endpoints array
+# index 0
+function add_endpoint_cfg_obj_setting() {
+    local setting=$1; shift
+    local value=$1; shift
+
+    do_jq --arg setting "${setting}" --arg value "${value}" '."endpoints"[0]."config"[0]."settings" += { ($setting): $value }'
+}
+
+#################################################################################
```
Comment (rafaelfolco) [inline crucible/from-file-json-manipulation-api/run.sh:133]: suggestions, not a hard requirement... 
functions above: this is the code, the api. Keep it in the main script. BTW, I don't like the term "run.sh"for all of it. I'd rename the api to run-file-api/update-template ...

and below this line: data replacements.... this is "per benchmark", so this is in the run.sh, or a prepare-run-file...

summary of suggestions:
1) api jq function in a separate file under run-file-api/update-template
2) jq calls in run.sh or prepare-run-file (latter implies in running prepare-run-file and then crucible run --from-file...)
3) data replacements abc=123 in a separated file that can be sourced from prepare-run-file

**Review by rafaelfolco — COMMENTED** [review]
this is pretty good btw, very nice api

File: crucible/from-file-json-manipulation-api/run.sh [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    fi
+}
+
+# change an existing run-params parameter value
+function override_run_param() {
+    local param=$1; shift
+    local value=$1; shift
+
+    do_jq --arg param "${param}" --arg value "${value}" '."run-params"[$param] = $value'
+}
```
Comment (k-rister) [inline crucible/from-file-json-manipulation-api/run.sh:51]: Different data types can require different syntax, so we may have to do things like this:

```
function override_run_param() {
    local type=$1; shift
    local param=$1; shift
    local value=$1; shift

    case "${type}" in
        "string")
            do_jq --arg param "${param}" --arg value "${value}" '."run-params"[$param] = $value'
            ;;
        "int")
            do_jq --arg param "${param}" --argjson value "${value}" '."run-params"[$param] = $value'
            ;;
    esac
}
```

**Pull request closed without merging**